theano/gof/type.py

```diff
@@ -497,6 +497,9 @@ def c_sync(self, name, sub):
     def c_code_cache_version(self):
         return (1,)
 
+    def may_share_memory(self, other):
+        return self is other
```

Comment:
I can fix the second problem.  

But to fix the first one would require a much more involved change and might still not be correct. I can envision some sort of attribute walker that checks recursively if attributes share memory (but that might be ok in some cases, so we'll get false positives).

Or we could change DebugMode to only try calling the function if it's there.  Which might actually be the better fix, since I'm pretty sure Generic is not the only Type without may_share_memory().